nodeadm/internal/containerd/oci_config.go

```diff
@@ -0,0 +1,148 @@
+package containerd
+
+import (
+	"errors"
+	"io/fs"
+	"os/exec"
+	"reflect"
+	"slices"
+	"strings"
+
+	"github.com/awslabs/amazon-eks-ami/nodeadm/internal/api"
+	"github.com/awslabs/amazon-eks-ami/nodeadm/internal/util"
+	"github.com/pelletier/go-toml/v2"
+	"go.uber.org/zap"
+)
+
+type instanceTypeMixin struct {
+	instanceFamilies []string
+	apply            func(*[]byte) error
+}
+
+func (m *instanceTypeMixin) matches(cfg *api.NodeConfig) bool {
+	instanceFamily := strings.Split(cfg.Status.Instance.Type, ".")[0]
+	return slices.Contains(m.instanceFamilies, instanceFamily)
+}
+
+var (
+	// TODO: fetch this list dynamically
+	nvidiaInstances         = []string{"p3", "p3dn", "p4d", "p4de", "p5", "g4", "g4dn", "g5", "g6"}
+	neuronInstances         = []string{"inf1", "inf2", "trn1", "trn1n"}
+	NvidiaInstanceTypeMixin = instanceTypeMixin{
+		instanceFamilies: nvidiaInstances,
+		apply:            applyNvidia,
+	}
+
+	NeuronInstanceTypeMixin = instanceTypeMixin{
+		instanceFamilies: neuronInstances,
+		apply:            applyNeuron,
+	}
+
+	mixins = []instanceTypeMixin{
+		NvidiaInstanceTypeMixin,
+		NeuronInstanceTypeMixin,
+	}
+)
+
+// applyInstanceTypeMixins adds the needed OCI hook options to containerd config.toml
```

Comment:
Comment here (and the name of the file) could be clearer. Is the intention of this interface to be a general-purpose way to do something special for certain instance types, or do we only want to handle alternative low-level runtimes?

Reply:
For now the intention is to only apply alternative low-level runtimes (if needed). I say if needed, because in the last commit I removed the special handling for Neuron, which Braynt commented was no longer needed. So for now it's just adding the nvidia runtime config.

Reply:
I think the instanceTypeMixin interface makes sense, just need a clearer file name